catchup/catchpointService.go

```diff
@@ -0,0 +1,534 @@
+			return cs.abort(fmt.Errorf("processStageLastestBlockDownload failed when calling StoreFirstBlock : %v", err))
+		}
+
+		err = cs.updateStage(ledger.CatchpointCatchupStateBlocksDownload)
+		if err != nil {
+			if attemptsCount <= maxBlockDownloadAttempts {
+				// try again.
+				blk = nil
+				continue
+			}
+			return cs.abort(fmt.Errorf("processStageLastestBlockDownload failed to update stage : %v", err))
+		}
+
+		// great ! everything is ready for next stage.
+		break
+	}
+	return nil
+}
+
+// processStageBlocksDownload is the fourth catchpoint catchup stage. It downloads all the reminder of the blocks, verifying each one of them against it's predecessor.
+func (cs *CatchpointCatchupService) processStageBlocksDownload() (err error) {
+	topBlock, err := cs.ledgerAccessor.EnsureFirstBlock(cs.ctx)
+	if err != nil {
+		return cs.abort(fmt.Errorf("processStageBlocksDownload failed, unable to ensure first block : %v", err))
+	}
+
+	// pick the lookback with the greater of either MaxTxnLife or MaxBalLookback
+	lookback := int(config.Consensus[topBlock.CurrentProtocol].MaxTxnLife)
+	if lookback < int(config.Consensus[topBlock.CurrentProtocol].MaxBalLookback) {
+		lookback = int(config.Consensus[topBlock.CurrentProtocol].MaxBalLookback)
+	}
+
+	cs.statsMu.Lock()
+	cs.stats.TotalBlocks = uint64(lookback)
+	cs.stats.AcquiredBlocks = 0
+	cs.stats.VerifiedBlocks = 0
+	cs.statsMu.Unlock()
+
+	prevBlock := &topBlock
+	fetcherFactory := MakeNetworkFetcherFactory(cs.net, 10, nil)
+	blocksFetched := 1 // we already got the first block in the previous step.
+	var blk *bookkeeping.Block
+	var client FetcherClient
+	for attemptsCount := 1; blocksFetched <= lookback; attemptsCount++ {
+		if err := cs.ctx.Err(); err != nil {
+			return cs.abort(err)
+		}
+
+		blk = nil
+		// check to see if the current ledger might have this block. If so, we should try this first instead of downloading anything.
+		if ledgerBlock, err := cs.ledger.Block(topBlock.Round() - basics.Round(blocksFetched)); err == nil {
+			blk = &ledgerBlock
+		} else {
+			switch err.(type) {
+			case ledger.ErrNoEntry:
+				// this is expected, ignore this one.
+			default:
+				cs.log.Warnf("processStageBlocksDownload encountered the following error when attempting to retrieve the block for round %d : %v", topBlock.Round()-basics.Round(blocksFetched), err)
+			}
+		}
+
+		if blk == nil {
+			fetcher := fetcherFactory.New()
+			blk, _, client, err = fetcher.FetchBlock(cs.ctx, topBlock.Round()-basics.Round(blocksFetched))
+			if err != nil {
+				if err == cs.ctx.Err() {
+					return cs.abort(err)
+				}
+				if attemptsCount <= maxBlockDownloadAttempts {
+					// try again.
+					continue
+				}
+				return cs.abort(fmt.Errorf("processStageBlocksDownload failed after multiple blocks download attempts"))
+			}
+			// success
+			client.Close()
+		}
+
+		cs.updateBlockRetrievalStatistics(1, 0)
+
+		// validate :
+		if prevBlock.BlockHeader.Branch != blk.Hash() {
+			// not identical, retry download.
```

Comment:
shouldn't all of these error conditions be checking `attemptsCount`? otherwise this can loop forever

Reply:
yes, fixed.